### PR TITLE
Add more nodes to elisp parser, remove `s` dependency, add vimscript

### PR DESCRIPTION
**Pull request by DevelopmentCool2449**
This commit includes some new features and changes:
* Add more nodes/rule set to elisp parser. This add support for folding lists and special forms such as `if`, `defvar` and etc.
* Remove dependency in `s.el` library. There was only 1 function from s.el in use, i replaced it by its built-in alternative.
* Add vimscript support through vimscript-ts-mode. It only add folding for `functions` and `if` conditionals.
* Add more missing `*-ts-modes` to `treesit-fold-range-alist` list. This also remove some modes that doesn't exists or unlikely to exist.

### CODE REVIEW

File: treesit-fold-parsers.el [hunk 1]
```diff
@@ -217,6 +217,8 @@
   "Rule set for Elisp."
   '((macro_definition    . treesit-fold-range-elisp-function)
     (function_definition . treesit-fold-range-elisp-function)
+    (special_form   . treesit-fold-range-seq)
+    (list           . treesit-fold-range-seq)
```
Comment (jcs090218) [inline treesit-fold-parsers.el:221]: `list` and `special_form` are almost everything, so I didn't include them in the first place. 🤔

File: treesit-fold.el [hunk 2]
```diff
@@ -391,14 +398,15 @@ This function is borrowed from `tree-sitter-node-at-point'."
            (end (cdr range))
            (ov (make-overlay beg end))
            (map (make-sparse-keymap)))
-      (keymap-set map "<mouse-1>" #'treesit-fold-open)
+      (define-key map [mouse-1] #'treesit-fold-open)
```
Comment (jcs090218) [inline treesit-fold.el:401]: Can you explain why the changes here? 🤔

Comment (DevelopmentCool2449) [reply treesit-fold.el:401]: It's only a little change that i thought would be a good idea include it, since define-key still used by most packages,
i'm reverting the changes if it's needed

File: treesit-fold-parsers.el (hunk 1 above)
Comment (DevelopmentCool2449) [reply treesit-fold-parsers.el:221]: as far as i could test, it doesn't conflict with `defun`s and `defmacro`s, 
It can be useful for fold large lists, or large conditionals

File: treesit-fold.el (hunk 2 above)
Comment (jcs090218) [reply treesit-fold.el]: No worries! I was just curious about the changes. 👍

File: treesit-fold-parsers.el (hunk 1 above)
Comment (jcs090218) [reply treesit-fold-parsers.el:221]: > as far as i could test, it doesn't conflict with defuns and defmacros,

Yes, you are right. :)

> It can be useful in some cases such as fold large lists, or large conditionals

My concern is most of the statements are list. Like `(require 'xxx)` is also identified as a list. 🤔

Comment (DevelopmentCool2449) [reply treesit-fold-parsers.el:221]: > > It can be useful in some cases such as fold large lists, or large conditionals
> 
> My concern is most of the statements are list. Like `(require 'xxx)` is also identified as a list. 🤔

it only would get folded if it's something like:
```elisp
(require 'subr-x
         nil
         t)
```
which i think it would be very rare that someone indent it like that.

Comment (jcs090218) [reply treesit-fold-parsers.el:221]: I guess we can try and see if this work. I am a bit worry on the indicator's performance. 🤔

Comment (jcs090218) [reply treesit-fold-parsers.el:221]: I think we still need some improvements, for example:

```elisp
(define-minor-mode cogru-mode
  "Minor mode `cogru-mode'."
  :global t
  :require 'cogru-mode
  :group 'cogru
  (if cogru-mode (cogru-mode--enable) (cogru-mode--disable)))
```

will fold into:

```elisp
(...)
```

It will be great if it can fold like `defun`, etc.

```elisp
(define-minor-mode cogru-mode
  ...)
```

`defvar`, `defcustom`, `defvar-local` all have the same issue. 🤔

Comment (DevelopmentCool2449) [reply treesit-fold-parsers.el:221]: > Feel free to split this into smaller PRs. It will be much easier for reviewers to review. :)

You are right, i will remove the elisp parser changes for a separated PR.

I apologize if i merged these commits in 1 PR, most of these changes would generate incompatibility each to other.

File: treesit-fold.el [hunk 3]
```diff
@@ -63,6 +62,7 @@
     (agda-mode              . ,(treesit-fold-parsers-agda))
     (arduino-mode           . ,(treesit-fold-parsers-arduino))
     (asm-mode               . ,(treesit-fold-parsers-asm))
+    (awk-mode               . ,(treesit-fold-parsers-c))
```
Comment (jcs090218) [inline treesit-fold.el:65]: Should we create a specific parser function for Awk? 🤔

Comment (DevelopmentCool2449) [reply treesit-fold.el:65]: as far as i know awk uses some of c syntax, such as `if`, `switch`, `for` and `while`.
so i don't think it should be necessary (for now)

Comment (jcs090218) [reply treesit-fold.el:65]: I prefer to wrap it like this: 🤔 

```elisp
(defun treesit-fold-parsers-awk ()
  "Rule set for Awk."
  (append (treesit-fold-parsers-c)))
```

And [tree-sitter-awk](https://github.com/Beaglefoot/tree-sitter-awk) doesn't seem to build on top of the tree-sitter-c. Maybe add a comment `;; TODO:`besides it. 🤔

Comment (DevelopmentCool2449) [reply treesit-fold.el]: > > It can be useful in some cases such as fold large lists, or large conditionals
> 
> My concern is most of the statements are list. Like `(require 'xxx)` is also identified as a list. 🤔

it only would get folded if it's something like:
```elisp
(require 'subr-x
         nil
         t)
```
which i think it would be very rare that someone indent it like that.

> I prefer to wrap it like this: 🤔
> 
> ```emacs-lisp
> (defun treesit-fold-parsers-awk ()
>   "Rule set for Awk."
>   (append (treesit-fold-parsers-c)))
> ```
> 
> And [tree-sitter-awk](https://github.com/Beaglefoot/tree-sitter-awk) doesn't seem to build on top of the tree-sitter-c. Maybe add a comment `;; TODO:`besides it. 🤔

Done

**Review by jcs090218 — APPROVED** [review]

**Comment by jcs090218** [comment]
LGTM! I appreciate your cooperation!

**Comment by DevelopmentCool2449** [comment]
> LGTM! I appreciate your cooperation!

Thank you, also i would like to know if there are plans for submit this package to nonGNU ELPA.
I have already some changes for allow the submition.

**Comment by jcs090218** [comment]
I never submitted a package to nonGNU ELPA since I already have my own ELPA, [JCS-ELPA](https://jcs-emacs.github.io/jcs-elpa/). Anyone can submit this package to non-GNU ELPA if they want to. :)

**Comment by DevelopmentCool2449** [comment]
> I never submitted a package to nonGNU ELPA since I already have my own ELPA, [JCS-ELPA](https://jcs-emacs.github.io/jcs-elpa/). Anyone can submit this package to non-GNU ELPA if they want to. :)

Glad to know, i've open a PR that is the last step to allow submiting it to NonGNU.
If there is no any problem also i would like to be the person to send a request for NonGNU submition.

**Comment by jcs090218** [comment]
> If there is no any problem also i would like to be the person to send a request for NonGNU submition.

Feel free to do so!
